### ISSUE TITLE
datasource (cloudwatch) errors too vague

### ISSUE THREAD

**Issue by Dieterbe** [grafana/grafana#16263]
In Grafana v6.0.1 (0c44a04) and using cloudwatch datasource, saw alerts fire with messages such

```
Execution Error: tsdb.HandleRequest() response error &{SerializationError: failed to decode query XML error response status code: 503, request id: caused by: EOF A <nil> [] []}"
```

and
```
Execution Error: tsdb.HandleRequest() response error &{RequestCanceled: request context canceled caused by: context deadline exceeded A <nil> [] []}"
```

These errors are too vague
1) no one knows what " tsdb.HandleRequest() " is.  it should be clearer and say something like  "datasource query error" or "error trying to query datasource". bonus points if including the type or name of the datasource
2) a more human friendly error message would help too, with fewer special characters.

I'm using cloudwatch as example here, but likely the same issue with other datasources.

**Comment by marefr** [comment]
Totally agree @Dieterbe that erro rmessages should be more human friendly. So given your examples we can probably do something like:

```
Failed to query datasource: &{SerializationError: failed to decode query XML error response status code: 503, request id: caused by: EOF A <nil> [] []}"
Failed to query datasource: &{RequestCanceled: request context canceled caused by: context deadline exceeded A <nil> [] []}"
```

Regarding the prefix `Execution Error:` that's something that Grafana UI is adding;
https://github.com/grafana/grafana/blob/206921d21b8c3acf71410c938259c06de3167cd9/public/app/features/alerting/state/reducers.ts#L19

Regarding adding the datasource type we could do that, but not sure since you probably already know what datasource type is used for your alert rule?

Our current suggestion for troubleshooting alerting is to inspect Grafana server log since the datasource in question may log more information. There's also this documentation: http://docs.grafana.org/alerting/rules/#troubleshooting about debug flags.

After thinking a bit of this, what I really would like us to do is to store structured information in addition to/instead of a single text message. 

We should be able to store a payload similar to what's shown in the query inspector so that you can inspect the actual query that was sent to the datasource. Additionally, one cool thing would be possibility to enable debug mode of an alert rule to be able to inspect both the actual query that was sent to the datasource and what the actual response was - this would be super useful for elasticsearch which has a really advanced query model where things easy can go wrong.

**Comment by Dieterbe** [comment]
> Regarding adding the datasource type we could do that, but not sure since you probably already know what datasource type is used for your alert rule?

if you go click on all the idvidual alert rules you can go find out yes. but when looking at the alert list and there's many of them in error state, it would be more convenient to have it in the overview.

> Our current suggestion for troubleshooting alerting is to inspect Grafana server log since the datasource in question may log more information. There's also this documentation: http://docs.grafana.org/alerting/rules/#troubleshooting about debug flags.

Grafanacloud customers don't have access to server logs, and they also cannot change the grafana configuration.

> We should be able to store a payload similar to what's shown in the query inspector so that you can inspect the actual query that was sent to the datasource. Additionally, one cool thing would be possibility to enable debug mode of an alert rule to be able to inspect both the actual query that was sent to the datasource and what the actual response was - this would be super useful for all datasources because it's really hard to debug alerting queries today since what you see in the UI is executing in the browser and alerting executing in the backend.

sounds good to me. Note that we would have to get this info about a query that failed in the past, the problem may not reproduce.

**Comment by marefr** [comment]
> if you go click on all the idvidual alert rules you can go find out yes. but when looking at the alert list and there's many of them in error state, it would be more convenient to have it in the overview.

True. Are you saying this as a user looking at someone else alert rules or you think this is generally true for your internal alerts as well?

> Grafanacloud customers don't have access to server logs, and they also cannot change the grafana configuration.

Very true. We don't have a good user story for this. Thinking out of the blue I think it would make more sense to make this available on grafana.com (admin interface) than in Grafana itself.

> Note that we would have to get this info about a query that failed in the past, the problem may not reproduce. 

If I understand you correctly you would like to be able to see sort of an audit trail of historic failed alert rules? Personally I think this is what the Grafana server log should be used for.

**Comment by Dieterbe** [comment]
> True. Are you saying this as a user looking at someone else alert rules or you think this is generally true for your internal alerts as well?

It doesn't really matter who set up the alerts originally. The scenario is as follows:
1) all is good in the world
2) a bunch of different alerts start firing
3) person looks at alert list and sees many different errors. at this point the person looking probably is getting stressed and is trying to figure out what is going wrong ASAP as they start thinking systems are down.
4) if the error messages were clearer, it would be obvious at this point that they are all firing because the same datasource is broken, and which one is broken.  Instead, currently, it's not immediately clear and they can either start diagnosing themselves (or also commonly: invoke the grafanacloud support team which is not a good way to spend our time)

> If I understand you correctly you would like to be able to see sort of an audit trail of historic failed alert rules? Personally I think this is what the Grafana server log should be used for.

not necessarily a trail of all failed queries. the point is simply that it should be clear from the messages that the datasource is broken, from where-ever the user is (e.g. alert list, or anywhere else they could be looking at the alert. maybe the notification itself). When people start panicing they're not gonna to grafana.com to change their grafana server settings

**Comment by marefr** [comment]
Thanks. 

Did a POC in #16286 for improving story for testing alert rules

**Comment by Dieterbe** [comment]
Not sure what the test rule functionality has to do with this

**Comment by marefr** [comment]
It's very related since it uses same code as real alert rule evaluation. But you're right the POC doesn't fix this issue. However as I wrote in PR I see the POC as good first step to explore what's possible to do for solving this issue moving forward.

**Comment by Bak3y** [comment]
Just ran into this, opsgenie and slack blowing up due to what appears to be grafana struggling to connect to prometheus, but relevant to the discussion the error messages are very unclear.

**Comment by armandgrillet** [comment]
The original issue concerned legacy alerting. The issue still exists in Grafana Alerting when the wrapping of errors made by datasources os not being done leading alerts to return abstract errors that do not describe the problem well. Checking the Grafana logs for alerting ones can help to debug these issues.
